rust/dns-over-tcp/src/client.rs

```diff
@@ -0,0 +1,417 @@
+                server,
+                &mut self.pending_queries_by_remote,
+                &mut self.sent_queries_by_remote,
+                &mut self.query_results,
+            );
+
+            let has_pending_dns_queries = !self
+                .pending_queries_by_remote
+                .entry(server)
+                .or_default()
+                .is_empty();
+
+            // Third, if the socket got closed, reconnect it.
+            if matches!(socket.state(), tcp::State::Closed) && has_pending_dns_queries {
+                let local_port = self
+                    .local_ports_by_socket
+                    .get(handle)
+                    .expect("must always have a port for each socket");
+                let local_endpoint = match server {
+                    SocketAddr::V4(_) => SocketAddr::new(ipv4_source.into(), *local_port),
+                    SocketAddr::V6(_) => SocketAddr::new(ipv6_source.into(), *local_port),
+                };
+
+                tracing::info!(local = %local_endpoint, remote = %server, "Re-connecting to DNS resolver");
+
+                socket
+                    .connect(self.interface.context(), server, local_endpoint)
+                    .expect(
+                        "re-connecting a closed socket with the same parameters should always work",
+                    );
+            }
+        }
+    }
+
+    fn sample_unique_ports(&mut self, num_ports: usize) -> Result<impl Iterator<Item = u16>> {
+        let mut ports = HashSet::with_capacity(num_ports);
+
+        if num_ports > self.port_range.len() {
+            bail!(
+                "Port range only provides {} values but we need {num_ports}",
+                self.port_range.len()
+            )
+        }
+
+        while ports.len() < num_ports {
+            ports.insert(self.rng.gen_range(self.port_range.clone()));
+        }
+
+        Ok(ports.into_iter())
+    }
+}
+
+fn send_pending_queries(
+    socket: &mut Socket,
+    server: SocketAddr,
+    pending_queries_by_remote: &mut HashMap<SocketAddr, VecDeque<Message<Vec<u8>>>>,
+    sent_queries_by_remote: &mut HashMap<SocketAddr, HashMap<u16, Message<Vec<u8>>>>,
+    query_results: &mut VecDeque<QueryResult>,
+) {
+    let pending_queries = pending_queries_by_remote.entry(server).or_default();
+    let sent_queries = sent_queries_by_remote.entry(server).or_default();
+
+    loop {
+        if !socket.can_send() {
+            break;
+        }
+
+        let Some(query) = pending_queries.pop_front() else {
+            break;
+        };
+
+        match codec::try_send(socket, query.for_slice_ref()).context("Failed to send DNS query") {
+            Ok(()) => {
+                let replaced = sent_queries.insert(query.header().id(), query).is_some();
+                debug_assert!(!replaced, "Query ID is not unique");
+            }
+            Err(e) => {
+                // We failed to send the query, declare the socket as failed.
+                socket.abort();
```

Comment:
Could it be that the socket's buffer is full but it can recover?

Reply:
If it is full, we should have not tried to write a message.